Simplify output algorithm

N1ckP3rsl3y opened the pull request:
Simplify the output algorithm to remove the need to access the "previous" output period and redundancy in calls to output information

- Remove use/mention of `tOffset`, `bFlush_output` and `SW_OUT_flush()` which add complexity to the output algorithm
- Update the output algorithm to collect/average values at the end of the each period rather than at the beginning of the next
- Update the documentation on the output algorithm to remove outdated information and to be up-to-date as of this PR

- Fix an unknown bug where establishment values had the chance to be reported one period before they should be (see commit and issue: 657be4e979869dbd2294c2dde4dd344972afaf61 & #494)
- close #494

N1ckP3rsl3y committed:
Update how new output periods are calculated

- When a new output period is achieved during simulation, it is the day immediately after the previous period was completed
- Following this, we then calculate which output periods to write out at the end of the day
- This approach - writing output after a period - forces us to reference the previous output period's information
- This adds some complexity to the calculation that isn't necessary
- An alternate idea is to determine if it's the end of the period on the last day, then store output on the same day

- Rename `newperiod` to `endperiod` in SW_MODEL_SIM
	* Modify the calculation for this in `SW_MDL_new_day`
		- Test for end of week, month and year (day is always true) by comparing the day of year to day for the end of each period

- New attribute in SW_OUT_RUN - `writeit`
	* Replaces the local version of `writeit` in `SW_OUT_write_today`
	* New function - `SW_OUT_new_day` to update this before the day starts
	* `begin_day` now calls this function after updating the model
	* No longer use `bFlush_output` in these calculations
- `SW_OUT_write_today` loses argument `bFlush_output`

N1ckP3rsl3y committed:
Remove instances of `tOffset` & `bFlush_output`

- The previous commit (1d4859cf6a3bf96ee797467fb6ecb96ca0d4f4a9) introduced the modification to collect values at the end of the output period
- When information is only handled in the current output period, there is no need for going to the "previous" one
- There is also no need for a specific call to write values after a year
	* Rather, the last day of the year should now produce the same as `SW_OUT_flush`

- Remove uses of `tOffset` and `bFlush_output` including within the interfaces for
	* `SW_OUT_sum_today`, `average_for`, `collect_values`, `SW_OUT_write_today` and `get_outstrleader`
- Remove function `SW_OUT_flush`
- `SW_OUT_sum_today` now collects values before averaging them

- These changes results in a slight modification to the example output for establishment day of year

N1ckP3rsl3y committed:
Gather "Today" values of `swcBulk` when averaging output

- When gathering `swcBulk` information, it can no longer be "Yesterday" as we need the "Today" value

- Update `average_for()` to gather the "Today" value of swcBulk

N1ckP3rsl3y committed:
Update output-related documentation in functions

N1ckP3rsl3y committed:
Update output algorithm documentation

- The version of the output algorithm contains outdated information that does not belong
- This description should be updated for the changes that are having on the branch

- Update the documentation to be up-to-date and also include a mention of SWNETCDF/SWNC/SWMPI modes when discussing `get_*_mem()`

N1ckP3rsl3y committed:
Fix `test_severe` segmentation fault

- `SW_MDL_new_day()` assumes the month is set (not the value held in "notime")
- The command "make clean test_severe" showed this can (and should) result in a segmentation fault

- Modify `SW_MDL_new_day()` to check if the current month is "notime", otherwise check if day of year is end of the current month

N1ckP3rsl3y committed:
Merge branch 'feature_trcoeq' into feature_simplify_output

N1ckP3rsl3y committed:
Update NEWS.md to mention establishment change

N1ckP3rsl3y committed:
Remove header for `SW_OUT_flush()`

dschlaep commented on src/SW_Output.c:
```diff
@@ -680,14 +675,14 @@ static void average_for(
-            curr_pd = (sw->ModelSim.week + 1) - tOffset;
-            div = (bFlush_output) ? sw->ModelSim.lastdoy % WKDAYS : WKDAYS;
+            curr_pd = sw->ModelSim.week + 1;
+            div = (doy == lastDoy) ? lastDoy % WKDAYS : WKDAYS;
```
In average_for(), the weekly `div` is calculated as `lastDoy % WKDAYS` if doy is the lastDoy. This seems to be wrong (this has been in the code for a very long time seemingly!): if lastDoy is the last day in a week then, then div = 0 (instead of 7 and it results in division by zero). Shouldn't this be `(lastDoy - 1) % WKDAYS + 1`?

For instance, setting `EndDoy  350` in "Inputs/domain.in" produces "Inf" for the week 50 in "Output/sw2_weekly.csv"

However, this branch is now producing monthly output only for a month if the simulation completed the entire month (`SW_MDL_new_day()`). I like that better than weekly output which is currently output even if a week is incomplete.

dschlaep committed:
Update NEWS

dschlaep committed:
Produce weekly and monthly output only for completed weeks and months

- Produce output for a week or month only if all days of that period have been simulated (only relevant if "EndDoy" from "Input/domain.in" is less than 365) or if a calendar year is completed
- Previously, output of partial weeks and months have been produced
- Previously, output of a week that was completed exactly on "EndDoy" (e.g., 350) resulted in division by zero (because div = 350 % 7 = 0)

dschlaep reviewed (approved):
This is a great PR! It's really a neat simplification. I added one comment about problems with output of weeks that are partial or when the simulation ends before the end of the calendar year. I addressed my comment directly with the commit 443f688591c2640f5054d4775a1d2a92f4701ac8 and approve of this PR.
